### PR TITLE
Use the pre-convention type mapping configuration in query

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Part of #25084

### CODE REVIEW

**Comment by AndriySvyryd** [comment]
@smitpatel @ajcvickers This change propagates the model to the query services by making them scoped. This worked better than I expected and won't break providers as long as they use the correct service registration pattern (our NTS plugins didn't).

However now I'm wondering whether this is what the users would expect to happen. They configure the mapping for `Properties()`, but we apply it for values that aren't mapped to any property, like constants and function returns. Perhaps we should introduce a separate method on `ModelConfigurationBuilder` just for this, something like `Scalars()`

File: src/EFCore/Infrastructure/EntityFrameworkServicesBuilder.cs [hunk 1]
```diff
@@ -95,9 +95,6 @@ public class EntityFrameworkServicesBuilder
                 { typeof(IMemoryCache), new ServiceCharacteristics(ServiceLifetime.Singleton) },
                 { typeof(IEvaluatableExpressionFilter), new ServiceCharacteristics(ServiceLifetime.Singleton) },
                 { typeof(IQueryTranslationPreprocessorFactory), new ServiceCharacteristics(ServiceLifetime.Singleton) },
-                { typeof(IQueryableMethodTranslatingExpressionVisitorFactory), new ServiceCharacteristics(ServiceLifetime.Singleton) },
```
Comment (smitpatel) [inline src/EFCore/Infrastructure/EntityFrameworkServicesBuilder.cs:98]: While we are here, should we also make `IQueryTranslationPreprocessorFactory` scoped too?

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 2]
```diff
@@ -21,7 +21,7 @@ namespace Microsoft.EntityFrameworkCore.Query
     public class RelationalQueryableMethodTranslatingExpressionVisitor : QueryableMethodTranslatingExpressionVisitor
     {
         private readonly RelationalSqlTranslatingExpressionVisitor _sqlTranslator;
-        private readonly WeakEntityExpandingExpressionVisitor _weakEntityExpandingExpressionVisitor;
+        private readonly SharedTypeEntityExpandingExpressionVisitor _sharedTypeEntityExpandingExpressionVisitor;
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:24]: Same exists for InMemory

**Comment by AndriySvyryd** [comment]
Also `Properties()` lets one to configure properties using the base type or generic definition, e.g. `Properties<MyWrappedType<>>()`. For scalars this is probably less useful and might be unexpected. It also adds complexity to type mapping finding code.

I propose that `Scalars<T>()` will only apply to `T` and `T?` if it's a value type and we will throw for `Scalars<T?>()` or if `T` is non-instantiable.

**Comment by ajcvickers** [comment]
I agree with the concerns over Properties, but I worry that figuring out when to use Properties and when to use Scalars will be hard. Even figuring what is and isn't a "scalar" can be hard. But I can also see that a single set of semantics (i.e. one method, however it is named) doesn't work very well. I wonder if we should punt on this for 6.0 and wait for feedback?

**Comment by AndriySvyryd** [comment]
> I agree with the concerns over Properties, but I worry that figuring out when to use Properties and when to use Scalars will be hard. Even figuring what is and isn't a "scalar" can be hard. But I can also see that a single set of semantics (i.e. one method, however it is named) doesn't work very well. 

This will probably require good documentation or perhaps just start with a warning that `Scalars` is not commonly used. In testing I found that it is rarely needed as we do a good job of inferring mappings, however in places where it is needed there are no workarounds except writing a type mapper plugin.

> I wonder if we should punt on this for 6.0 and wait for feedback?

Or we can ship it in RC1 and wait for feedback.

File: src/EFCore.Cosmos/Query/Internal/InExpression.cs [hunk 3]
```diff
@@ -27,7 +26,7 @@ public class InExpression : SqlExpression
             SqlExpression item,
             bool negated,
             SqlExpression values,
-            CoreTypeMapping typeMapping)
+            CoreTypeMapping? typeMapping)
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Query/Internal/InExpression.cs:29]: This cannot be null.

**Comment by ajcvickers** [comment]
Okay with me to ship in RC1.

**Review by smitpatel — APPROVED** [review]

File: src/EFCore.Cosmos/Extensions/CosmosServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -128,10 +126,10 @@ public static IServiceCollection AddEntityFrameworkCosmos(this IServiceCollectio
                     b => b
                         .TryAddSingleton<ICosmosSingletonOptions, CosmosSingletonOptions>()
                         .TryAddSingleton<ISingletonCosmosClientWrapper, SingletonCosmosClientWrapper>()
-                        .TryAddSingleton<ISqlExpressionFactory, SqlExpressionFactory>()
                         .TryAddSingleton<IQuerySqlGeneratorFactory, QuerySqlGeneratorFactory>()
-                        .TryAddSingleton<IMethodCallTranslatorProvider, CosmosMethodCallTranslatorProvider>()
-                        .TryAddSingleton<IMemberTranslatorProvider, CosmosMemberTranslatorProvider>()
+                        .TryAddScoped<ISqlExpressionFactory, SqlExpressionFactory>()
+                        .TryAddScoped<IMemberTranslatorProvider, CosmosMemberTranslatorProvider>()
+                        .TryAddScoped<IMethodCallTranslatorProvider, CosmosMethodCallTranslatorProvider>()
```
Comment (roji) [inline src/EFCore.Cosmos/Extensions/CosmosServiceCollectionExtensions.cs:132]: Should RelationalMethodCallTranslatorProviderDependencies and RelationalMemberTranslatorProviderDependencies also be modified to contain the Model, assuming that translators need to look up mappings by type?

**Comment by roji** [comment]
> This change propagates the model to the query services by making them scoped. This worked better than I expected and won't break providers as long as they use the correct service registration pattern (our NTS plugins didn't).

Just spent a few hours tracking down some seemingly unrelated test failures to this...

**Comment by AndriySvyryd** [comment]
> Just spent a few hours tracking down some seemingly unrelated test failures to this...

Any suggestions for other providers? Should we emphasize that `TryAddProviderSpecificServices` should only be used for provider-specific services?

**Comment by roji** [comment]
Yeah... In theory would it be possible to actually check if services registered via TryAddProviderSpecificServices are actually not provider-specific, and throw an informative message? Though if that's complicated it may not be worth it...

**Pull request merged**